orttraining/orttraining/python/orttraining_pybind_state.cc

```diff
@@ -357,16 +362,106 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
       .def("reserve", [](std::vector<OrtValue>* v, const size_t len) { v->reserve(len); })
       .def("shrink_to_fit", [](std::vector<OrtValue>* v) { v->shrink_to_fit(); })
       .def("__len__", [](const std::vector<OrtValue>& v) { return v.size(); })
-      .def("__iter__", [](const std::vector<OrtValue>& v) {
-        return py::make_iterator(v.cbegin(), v.cend());
-      },
-           py::keep_alive<0, 1>())
+      .def(
+          "__iter__", [](const std::vector<OrtValue>& v) {
+            return py::make_iterator(v.cbegin(), v.cend());
+          },
+          py::keep_alive<0, 1>())
       .def("__getitem__", [](const std::vector<OrtValue>& v, const size_t idx) {
         return v.at(idx);
       })
       .def("dlpack_at", [](std::vector<OrtValue>* v, const size_t idx) {
         return py::reinterpret_steal<py::object>(ToDlpack(v->at(idx)));
-      });
+      })
+      .def("to_dlpack", [](const std::vector<OrtValue>& v, py::object to_tensor) -> py::list {
+
+        if (v.size() == 0)
+          return py::list();
+
+        py::list res;
+
+        PyObject* capsule;
+        PyObject* handle = to_tensor.ptr();
+        PyObject* obj;
+
+        py::gil_scoped_acquire acquire;
+
+        const Tensor& tensor = v[0].Get<Tensor>();
+
+        if (tensor.Location().device.Type() == OrtDevice::CPU) {
+          OrtDLManagedTensor ort_dlmanaged_tensor;
+          capsule = PyCapsule_New(&ort_dlmanaged_tensor.tensor, "dltensor", NULL);
+          if (capsule == NULL)
+            throw std::runtime_error("Empty capsule returned.");
+
+          for (auto it : v) {
+            // The same capsule is reused but FromDLPack rename the capsule into used_dltensor.
+            PyCapsule_SetName(capsule, "dltensor");
+            dlpack::OrtValueToDlpack(it, (void*)&ort_dlmanaged_tensor);
+    #if (PY_MAJOR_VERSION >= 3 && PY_MINOR_VERSION >= 9)
+            obj = PyObject_CallOneArg(handle, capsule);
+    #else
+            obj = PyObject_CallFunctionObjArgs(handle, capsule, NULL);
+    #endif
+            if (obj == NULL)
+              throw std::runtime_error("Empty tensor returned.");
+            res.append(py::handle(obj));
+          }
+        } else {
+          // This piece of code should be removed when the issue on GPU is found.
```

Comment:
What kind of issue here?

Reply:
When an OrtValue is created, it copied whatever needs to be copied but the deleter attached the ortvalue deletes the data pointer and the DLPack structure used to create it even if it is not needed anymore. I first assumed the python capsule was responsible for that but it only is when the capsule is not consumed. Pytorch does the same thing.

```
OrtValue DlpackToOrtValue(DLManagedTensor* dlpack, bool is_bool_tensor) {
  // ORT only supports contiguous tensor for now.
  ORT_ENFORCE(IsContiguousTensor(dlpack->dl_tensor), "ORT only supports contiguous tensor for now.");
  OrtDevice device = GetOrtDevice(dlpack->dl_tensor.device);
  MLDataType data_type = GetOrtValueDataType(dlpack->dl_tensor.dtype, is_bool_tensor);
  OrtMemoryInfo info(GetOrtDeviceName(device), OrtDeviceAllocator, device, device.Id());
  std::unique_ptr<Tensor> p_tensor = std::make_unique<Tensor>(
      data_type, TensorShape(dlpack->dl_tensor.shape, static_cast<size_t>(dlpack->dl_tensor.ndim)),
      dlpack->dl_tensor.data, info);

  OrtValue ort_value;
  std::function<void(void*)> deleter = [dlpack](void* p) {
    if (dlpack->deleter != NULL)
      dlpack->deleter(dlpack);
    auto deleter = DataTypeImpl::GetType<Tensor>()->GetDeleteFunc();
    if (deleter != NULL)
      deleter(p);
  };

  ort_value.Init(p_tensor.release(), DataTypeImpl::GetType<Tensor>(), deleter);
  return ort_value;
}
```


```
void DlpackCapsuleDestructor(PyObject* data) {
  DLManagedTensor* dlmanaged_tensor = reinterpret_cast<DLManagedTensor*>(PyCapsule_GetPointer(data, "dltensor"));
  if (dlmanaged_tensor) {
    // The dlmanaged_tensor has not been consumed, call deleter ourselves.
    dlmanaged_tensor->deleter(const_cast<DLManagedTensor*>(dlmanaged_tensor));
  } else {
    // The dlmanaged_tensor has been consumed,
    // PyCapsule_GetPointer has set an error indicator.
    PyErr_Clear();
  }
}
```